### PR TITLE
Support lambda with a trick

### PR DESCRIPTION
**Pull request by NoodlesPi**
I have a naive but helpful idea to sovle this issue. I notice that when the parameter in MethodSpec.methodBuilder is a empty string "", it will throw a java.lang.NullPointerException. Actually, we can use this feature to build our lambda because we can stipulate that a method with empty string here is the lambda. To realize it, we just need some slightly revise which do not affect previous code.
Example:
```
MethodSpec lambda = MethodSpec.methodBuilder("")
                        .addParameter(String.class, "a")
                        .addParameter(String.class, "b")
                        .addStatement("return $N.length() - $N.length()", "a", "b")
                        .build();

        TypeSpec helloWorld = TypeSpec.classBuilder("HelloWorld")
                .addMethod(MethodSpec.methodBuilder("sortByLength")
                        .addParameter(ParameterizedTypeName.get(List.class, String.class), "strings")
                        .addStatement("$T.sort($N, $L)", Collections.class, "strings", lambda)
                        .build())
                .build();
```
Output with my modification:
```
class HelloWorld {
  void sortByLength(List<String> strings) {
    Collections.sort(strings, (a, b) -> {
          return a.length() - b.length();
        });
  }
}
```

### CODE REVIEW

File: src/main/java/com/squareup/javapoet/ParameterSpec.java [hunk 1]
```diff
@@ -151,6 +177,7 @@ public Builder addJavadoc(CodeBlock block) {
     public Builder addAnnotations(Iterable<AnnotationSpec> annotationSpecs) {
       checkArgument(annotationSpecs != null, "annotationSpecs == null");
       for (AnnotationSpec annotationSpec : annotationSpecs) {
+        System.out.println(annotationSpec.type);
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ParameterSpec.java:180]: revert

File: src/main/java/com/squareup/javapoet/ParameterSpec.java [hunk 2]
```diff
@@ -37,28 +38,40 @@
...
+    }
     if (varargs) {
       TypeName.asArray(type).emit(codeWriter, true);
     } else {
-      type.emit(codeWriter);
+      if (type != null)
+        type.emit(codeWriter);
     }
-    codeWriter.emit(" $L", name);
+    if (type != null)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ParameterSpec.java:71]: curly braces

File: src/main/java/com/squareup/javapoet/ParameterSpec.java [hunk 3]
```diff
@@ -37,28 +38,40 @@
...
-    codeWriter.emitModifiers(modifiers);
+    if (type != null) {
+      codeWriter.emitAnnotations(annotations, true);
+      codeWriter.emitModifiers(modifiers);
+    }
     if (varargs) {
       TypeName.asArray(type).emit(codeWriter, true);
     } else {
-      type.emit(codeWriter);
+      if (type != null)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ParameterSpec.java:68]: curly braces

File: src/main/java/com/squareup/javapoet/ParameterSpec.java [hunk 4]
```diff
@@ -37,28 +38,40 @@
...
   public final CodeBlock javadoc;
+  public final boolean isLambda;
 
   private ParameterSpec(Builder builder) {
     this.name = checkNotNull(builder.name, "name == null");
     this.annotations = Util.immutableList(builder.annotations);
     this.modifiers = Util.immutableSet(builder.modifiers);
-    this.type = checkNotNull(builder.type, "type == null");
+    this.isLambda = builder.isLambda;
+    if (!isLambda)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ParameterSpec.java:48]: curly braces

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 5]
```diff
@@ -402,6 +427,8 @@ public Builder addParameter(ParameterSpec parameterSpec) {
     }
 
     public Builder addParameter(TypeName type, String name, Modifier... modifiers) {
+      if (isLambda)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java:430]: curly braces

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 6]
```diff
@@ -389,7 +413,8 @@ public Builder returns(Type returnType) {
     }
 
     public Builder addParameters(Iterable<ParameterSpec> parameterSpecs) {
-      checkArgument(parameterSpecs != null, "parameterSpecs == null");
+      if (!isLambda)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java:416]: curlies

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 7]
```diff
@@ -188,6 +200,7 @@ public static Builder methodBuilder(String name) {
     return new Builder(name);
   }
 
+
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java:203]: revert

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 8]
```diff
@@ -134,8 +144,10 @@ void emit(CodeWriter codeWriter, String enclosingName, Set<Modifier> implicitMod
       codeWriter.indent();
       codeWriter.emit(code, true);
       codeWriter.unindent();
-
-      codeWriter.emit("}\n");
+      if (isLambda)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java:147]: curly braces

File: src/main/java/com/squareup/javapoet/MethodSpec.java [hunk 9]
```diff
@@ -94,7 +99,10 @@ void emit(CodeWriter codeWriter, String enclosingName, Set<Modifier> implicitMod
     if (isConstructor()) {
       codeWriter.emit("$L($Z", enclosingName);
     } else {
-      codeWriter.emit("$T $L($Z", returnType, name);
+      if (isLambda)
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/MethodSpec.java:102]: curly braces

File: src/main/java/com/squareup/javapoet/ParameterSpec.java [hunk 10]
```diff
@@ -37,28 +38,40 @@
...
+      this.type = checkNotNull(builder.type, "type == null");
+    else
+      this.type = null;
     this.javadoc = builder.javadoc.build();
+
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ParameterSpec.java:53]: remove

File: src/main/java/com/squareup/javapoet/ParameterSpec.java [hunk 11]
```diff
@@ -86,10 +99,14 @@ void emit(CodeWriter codeWriter, boolean varargs) throws IOException {
   public static ParameterSpec get(VariableElement element) {
     checkArgument(element.getKind().equals(ElementKind.PARAMETER), "element is not a parameter");
 
+    // Copy over any annotations from element.
+    List<AnnotationSpec> annotations = element.getAnnotationMirrors()
+        .stream()
+        .map((mirror) -> AnnotationSpec.get(mirror))
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ParameterSpec.java:105]: nit;
```suggestion
        .map(AnnotationSpec::get)
```

File: src/main/java/com/squareup/javapoet/TypeName.java [hunk 12]
```diff
@@ -38,6 +38,7 @@
 import javax.lang.model.type.TypeMirror;
 import javax.lang.model.util.SimpleTypeVisitor8;
 
+
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/TypeName.java:41]: revert

**Comment by NoodlesPi** [comment]
> I'm tempted to instead force the use of an anonymous `ClassSpec` which could be converted to a lambda given a source level. This change lacks a ton of tests. I would need to see coverage of edge cases like inline lambdas `e -> e.foo()`, block lambdas `e -> { return e.foo(); }`, typed argument-lambdas like `String e -> e.foo()`, annotated argument lambdas like `@NonEmpty var e -> e.foo()`, and annotated typed lambdas like `@NonEmpty String e -> e.foo()` at least.

Yes, sure. But one issue with anonymous classes is that if the implementation of your anonymous class is very simple, such as an interface that contains only one method, then the syntax of anonymous classes may seem unwieldy and unclear. In these cases, you're usually trying to pass functionality as an argument to another method. This is the intention of the lambda. And from the user's perspective, they just care the functionality when use lambda. Therefore, it's may more reasonable to implements lambda using 'MethodSpec'. As for anonymous 'ClassSpec', it's also can be converted to a lambda but it seems not straightforward.
I just present my ideas here and do a simple implementation without enough tests. If you think this idea is OK, i will complete my work.
